patch/driver-support-optoe-write-max.patch

```diff
@@ -0,0 +1,86 @@
+From ca9daf437f0910493a6eb8ced900662223ea3cf7 Mon Sep 17 00:00:00 2001
+From: Prince George <prgeor@microsoft.com>
+Date: Mon, 16 Aug 2021 16:04:15 +0000
+Subject: [PATCH] Dynamic write_max support for optoe driver  
+ In current optoe driver, the value of write_max is  hardcoded to one byte.  
+ CMIS spec supports firmware upgrade on QSFP-DD transceivers which means  
+ there is significant overhead due to this one byte write limit when a
+ platform wants to upgrade the firmware over an i2c bus at 100Khz. 
+ The  overhead is more pronounced when a platform has shared master and 
+ platform needs to perform firmware upgrade  on multiple transceivers.
+
+ Added sysfs file 'write_max' for each i2c device (that use optoe driver)
+ so that the platform can override the default write max which is one byte.
+
+ Test result:-
+ Firmware file of size 1.9MB takes ~ 33 mins with write-max size = 1 Byte
+ Firmware file of size 1.9MB takes ~ 8 mins with write-max size = 64 Bytes
+
+Signed-off-by: Prince George <prgeor@microsoft.com>
+---
+ drivers/misc/eeprom/optoe.c | 36 ++++++++++++++++++++++++++++++++++++
+ 1 file changed, 36 insertions(+)
+
+diff --git a/drivers/misc/eeprom/optoe.c b/drivers/misc/eeprom/optoe.c
+index 16287fdc5..cacfb1848 100644
+--- a/drivers/misc/eeprom/optoe.c
++++ b/drivers/misc/eeprom/optoe.c
+@@ -822,6 +822,39 @@ static int optoe_remove(struct i2c_client *client)
+ 	return 0;
+ }
+ 
++static ssize_t show_dev_write_max_size(struct device *dev,
++			struct device_attribute *dattr, char *buf)
++{
++	struct i2c_client *client = to_i2c_client(dev);
++	struct optoe_data *optoe = i2c_get_clientdata(client);
++	ssize_t count;
++
++	mutex_lock(&optoe->lock);
++	count = sprintf(buf, "%u\n", optoe->write_max);
```

Comment:
I believe you want to use `scnprintf` here instead of `sprintf`.
Even though formatting a unsigned integer is unlikely to reach the `PAGE_SIZE` limit of `buf` it has the advantage of returning the size of the written buffer minus the `\0` char.
I believe this is the best practice since it's part of this documentation https://www.kernel.org/doc/html/latest/filesystems/sysfs.html#reading-writing-attribute-data

Reply:
The (*show) API is not very particular about excluding the null char from the returned count. The preference of scnprintf() over sprintf() is more from overflow protection mentioned here:-  
[https://elixir.bootlin.com/linux/v5.9.16/source/Documentation/filesystems/sysfs.rst#L247]

I think we are pretty much safe here to use sprintf().

Reply:
[https://elixir.bootlin.com/linux/v5.9.16/source/Documentation/filesystems/sysfs.rst#L247](https://elixir.bootlin.com/linux/v5.9.16/source/Documentation/filesystems/sysfs.rst#L247)